### PR TITLE
migrate to dune + opam file + alcotest

### PR DESCRIPTION
**Pull request by c-cube**
I must say, I didn't migrate everything in the makefile (esp. regarding documentation)

### CODE REVIEW

File: src/dune [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+(library
+  (name calendarLib)
+  (public_name calendar)
+  (wrapped true)
```
Comment (anuragsoni) [inline src/dune]: isn't this the default behaviour for dune?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+_build
+*.install
+.merlin
```
Comment (anuragsoni) [inline .gitignore]: maybe add `_opam` as well, for people using local switches?

**Comment by c-cube** [comment]
Note: the tests are pretty broken, they rely on global side effects and this PR changes their order. I think it's better to wait for #7 to fix them.

**Comment by pmetzger** [comment]
Which order do you want the pulls committed in? 7 first then this?

**Comment by c-cube** [comment]
Maybe I can merge them into one single PR…

**Comment by pmetzger** [comment]
That might be a good idea.

**Comment by pmetzger** [comment]
BTW, I'm tempted to add you to `CODEOWNERS`, @c-cube. How do you feel about that?

**Comment by c-cube** [comment]
Ok, let me merge the PRs into one. Regarding codeowners it's probably not necessary, I'm not sure I'll have a lot of motivation in the future :-s

**Comment by c-cube** [comment]
*sigh*. should we use the docker-based travis script? it's much faster.


Something like:
```
language: c
install: wget https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.travis-docker.sh
script: bash -ex .travis-docker.sh
services:
- docker
env:
  global:
  - PINS="calendar:."
  - DISTRO="ubuntu-16.04"
  matrix:
  - PACKAGE="calendar" OCAML_VERSION="4.03.0" DEPOPTS=""
  - PACKAGE="calendar" OCAML_VERSION="4.04.2" DEPOPTS=""
  - PACKAGE="calendar" OCAML_VERSION="4.05.0" DEPOPTS=""
  - PACKAGE="calendar" OCAML_VERSION="4.06.0" DEPOPTS=""
```

**Comment by loxs** [comment]
I don't see why not use docker.

**Comment by loxs** [comment]
Also, documentation seems to be pretty broken right now. Odoc doesn't generate anything when calendar is installed via Opam, so not migrating the documentation part of the Makefile seems fine to me, especially as I plan a major doc rehaul for 3.0

**Comment by pmetzger** [comment]
Shouldn't this be against `master`? Why is it against this "3.x" thing?

File: .travis.yml [hunk 3]
```diff
@@ -1,16 +1,14 @@
...
   matrix:
-  - OCAML_VERSION=4.03
-  - OCAML_VERSION=4.04
-  - OCAML_VERSION=4.05
-  - OCAML_VERSION=4.06
-  - OCAML_VERSION=4.07
-os:
-  - linux
-  - osx
+  - PACKAGE="calendar" OCAML_VERSION="4.03.0" TESTS=1 EXTRA_DEPS="alcotest"
```
Comment (anuragsoni) [inline .travis.yml]: from what i remember the new docker containers for `opam2` do not use the minor version for ocaml switch names anymore. https://github.com/ocaml/infrastructure/wiki/Containers#selecting-a-specific-compiler

**Comment by c-cube** [comment]
Locally documentation builds properly for me (with warnings though). I think you have to ask opam to build doc explicitly?

**Comment by c-cube** [comment]
hurray for the successful build? :grin:

File: Makefile [hunk 4]
```diff
@@ -1,295 +1,16 @@
...
-# Depend
-########
+	@dune clean
 
-.depend depend: $(GENERATED)
-	rm -f .depend
-	$(CAMLDEP) -I src -I tests src/*.ml src/*.mli tests/*.ml tests/*.mli \
-	  > .depend
+test:
+	@exec dune runtest --no-buffer --force
```
Comment (rizo) [inline Makefile]: Out of curiosity why `exec`?

File: Makefile [hunk 5]
```diff
@@ -1,295 +1,16 @@
...
 
-view-depend:
-	$(CAMLDOT) .depend | dot -Tps | gv -
+doc:
+	@dune build @doc
 
-include .depend
+.PHONY: all clean test
```
Comment (rizo) [inline Makefile]: `doc`?

File: README.md [hunk 6]
```diff
@@ -26,13 +25,10 @@ Older OCaml versions are unsupported.
...
 - `LGPL`		  Information about LGPL
 - `README.md`		  This file
-- `Makefile.in`	  Makefile used by configure
-- `configure`	  Script generating Makefile
-- `configure.in`	  Script generating configure (with autoconf)
 - `calendar_faq.txt`  FAQ frow which some algorithms come
 - `doc`		  HTML documentation of the API
 - `src`		  Source files directory
-- `target`		  Directory containing the built library
+- `_build/default/`		  Directory containing the built library
```
Comment (rizo) [inline README.md]: A nit: don't use a tab character here as it's not rendered anyway.

File: .gitignore (hunk 2 above)
Comment (rizo) [reply .gitignore]: Aren't local switches typically used in user projects? Will calendar itself ever have a local switch?

**Comment by c-cube** [comment]
This PR is big enough already, I'd say? Maybe you can base your branch on this one

File: README.md (hunk 6 above)
Comment (c-cube) [reply README.md]: it's how it was before…

File: .gitignore (hunk 2 above)
Comment (anuragsoni) [reply .gitignore]: That's true. Maybe there's a nicer way to do things but sometimes i like to just go `opam switch create . <ocaml-version>` when trying out a new project so it installs everything i need to explore without messing with the my system wide switch. Its very useful when a given project has constraints about compiler version/libraries that don't play nice with one's global switches. But that being said, its fairly straightforward to add `_opam` in the `.git/info/exclude` so its probably not important to add it here as well.

Comment (anuragsoni) [reply .gitignore]: One reason that I noticed this was also ignoring these files in any other tool like `Ctrl-P`, `fzf`, `ripgrep` etc which respect the ignore files.

**Comment by rizo** [comment]
I'll wait until this is merged. They're just warnings after all.

Thanks for the changes! :)

File: .gitignore (hunk 2 above)
Comment (rizo) [reply .gitignore]: That's a cool trick. You can also add `_opam` to your global `.gitignore`, I think those tools respect it.

**Comment by loxs** [comment]
Also maybe @rizo's changes to documentation could also go into 2.05

**Comment by c-cube** [comment]
so should I rebase + modify the deps? Personally I think we should merge as is, it's wasteful to undo changes for 2.x just to redo them later. Who is going to want to upgrade calendar but wouldn't be ready to switch to `re.str`?

**Comment by loxs** [comment]
Hmm, true, forgot about the deps change.
Another possibility would be to just call this 2.0 and merge as is.

**Comment by loxs** [comment]
The more I think about it, trying to make the old version pretty (2.04) is probably not that important. We can polish this a bit more and release 3.0, then if we need to do another breaking change (maybe if we remove the shared state), we can even bump up the major version again if deemed necessary.

**Pull request merged**